### DISCUSSION TITLE
Extend timestamp device class

### DISCUSSION THREAD

**Discussion by KapJI** [home-assistant/architecture#635] [Entity Models]
Currently it only supports `unavailable` and `unknown` states apart from an actual timestamp. These states are usually used for errors. `unavailable` state for example causes entity marked with a red sign in the entity manager.

Some sensor entities like alarms or timers might want to have some empty value instead of timestamp which is not an error when timestamp is not set.

The only viable option right now is `unknown` state but it's confusing for people and doesn't correctly reflect the state because it's actually known.

In https://github.com/home-assistant/frontend/pull/10006 I proposed to add a new `not_set` state which can solve this issue and provide a meaningful message in the frontend. Other options could be `none` (as a string, `None` in Python will become `unknown`, this variant is a bit confusing for developers) or `off`.

The alternative approach is in https://github.com/home-assistant/frontend/pull/10016. When state is not a timestamp instead of rendering `Invalid timestamp` message, pass state to `computeStateDisplay`. This will allow arbitrary non timestamp states and may hide errors when the state is actually a timestamp in invalid format. It doesn't add any new fixes states but it's probably too permissive.

Another alternative I have considered is changing the device class dynamically based on whether the timestamp is set or not. This works on the current version of HA but I think changing device class dynamically is a bad practice and should be avoided.

One more option is using a binary sensor with `on` and `off` states and save timestamp in state attributes. Although it works this leads to worse UX for users because from the alarm sensor they want to see in the frontend when it'll ring if it's set.

**Comment by ludeeus** [comment]
We should not pollute the device class state with arbitrary strings.
The main driver for this seems to be a frontend concern. If something needs to change IMO it's better to have the frontend render "None" instead of "Unknown" for `None`.

**Reply by KapJI** [reply]
Yes, the main concern for me is the frontend. Although conversion from `None` to `"unknown"` happens in the core and rendering `"unknown"` state as "None" in the frontend doesn't seem right. Even if it's just for timestamp device class.

**Reply by jlsjonas** [reply]
I think there's a valid case for "none" vs "unknown"; for any timestamp class related to events; "unknown" implies it's... _unknown_, while there might _actively_ be no upcoming timestamps at any given time.

I do fully agree that we shouldn't _pollute_ classes with just arbitrary strings; but I don't think _none_ is that arbitrary for a lot of use-cases

Just my 2 cents...

**Comment by davidrustingha** [comment]
Not to put people on a rush, but I'm interested to see where we stand on this matter.

**Comment by joeblack2k** [comment]


**Reply by frenck** [reply]
That is an odd question, as you can see all the history above. The answer is thus: no updates after Feb 8, 2022.

**Comment by EuleMitKeule** [comment]
I strongly support this change since upcoming timestamp sensors need to have a state for when there is no upcoming timestamp that doesn't indicate an error.
